runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogDataSerializerUtils.java

```diff
@@ -0,0 +1,211 @@
+                        buffer.writeByte(map.size());
+                        map.forEach((key, value) -> {
+                            serialize(buffer, key);
+                            serialize(buffer, value);
+                        });
+                    })
+                    .put(Map.class, (buffer, payload) -> {
+                        Map<?, ?> map = (Map<?, ?>) payload;
+                        buffer.writeInt(map.size());
+                        map.forEach((key, value) -> {
+                            serialize(buffer, key);
+                            serialize(buffer, value);
+                        });
+                    })
+                    .put(ByteBuf.class, (buffer, payload) -> {
+                        ByteBuf b = ((ByteBuf) payload).slice();
+                        b.resetReaderIndex();
+                        int bytes = b.readableBytes();
+                        buffer.writeInt(bytes);
+                        buffer.writeBytes(b, bytes);
+                    })
+                    .put(IMetadata.DataRank.class, (buffer, payload) -> {
+                        serialize(buffer, ((IMetadata.DataRank) payload).getRank());
+                        serialize(buffer, ((IMetadata.DataRank) payload).getUuid());
+                    })
+                    .put(IMetadata.LogUnitMetadataType.class, (buffer, payload) ->
+                            serialize(buffer, ((IMetadata.LogUnitMetadataType) payload).asByte()))
+                    .put(CheckpointEntry.CheckpointEntryType.class, (buffer, payload) ->
+                            serialize(buffer, ((CheckpointEntry.CheckpointEntryType) payload).asByte()))
+                    .put(Codec.Type.class, (buffer, payload) -> buffer.writeInt(((Codec.Type) payload).getId()))
+                    .build()
+    );
+
+    static ByteBuf byteBufFromBuffer(byte[] data) {
+        ByteBuf buffer = Unpooled.wrappedBuffer(data);
+        return fromBuffer(buffer, ByteBuf.class);
+    }
+
+    /**
+     * Build payload from Buffer
+     *
+     * @param buf The buffer to deserialize.
+     * @param cls The class of the payload.
+     * @param <T> The type of the payload.
+     * @return payload
+     */
+    @SuppressWarnings("unchecked")
+    static <T> T fromBuffer(ByteBuf buf, Class<T> cls) {
+        if (constructorMap.containsKey(cls)) {
+            return (T) constructorMap.get(cls).construct(buf);
+        } else if (cls.isEnum()) {
+            // We only know how to deal with enums with a type map
+            try {
+                Map<Byte, T> typeMap = (Map<Byte, T>) cls.getDeclaredField("typeMap").get(null);
+                constructorMap.put(cls, x -> typeMap.get(x.readByte()));
+                return (T) constructorMap.get(cls).construct(buf);
+            } catch (NoSuchFieldException e) {
+                throw new RuntimeException("Only enums with a typeMap are supported!");
+            } catch (IllegalAccessException e) {
```

Comment:
All this reflection is expensive to do per LogData. I think we need to revisit this later and just inline the metadata into the protobuf type.